src/reducers/transactionReducer.ts

```diff
@@ -45,7 +45,29 @@ const createTransaction = (state: TransactionState, initialTransaction: Transact
 export default function transactionReducer(state: TransactionState, action: TransactionsActionType): TransactionState {
   switch (action.type) {
     case TransactionActionTypes.SET_ESTIMATED_FEE:
-      return { ...state, estimatedFee: action.payload.estimatedFee };
+      return {
+        ...state,
+        estimatedFee: action.payload.estimatedFeeError ? null : action.payload.estimatedFee,
+        estimatedFeeError: action.payload.estimatedFeeError
+      };
+    case TransactionActionTypes.RESET_ESTIMATED_FEE:
+      return {
+        ...state,
+        estimatedFee: null,
+        estimatedFeeError: null
+      };
```

Comment:
Couldn't this be done via `setEstimateFee(null, null)`? Like we can have a separate action creator for that but we don't need a separate action and reducer branch.

Reply:
That's said, it's imho already much, much better than independent parameters. We could keep this action (but renamed), because what I feel is missing is the `isEstimatedFeeLoading` indicator. Now you don't really know if the `fee` is `null` because it's supposed to be `null` (due to some other things missing) or if it's actually loading.

Reply:
Please check after this commit: https://github.com/paritytech/parity-bridges-ui/pull/186/commits/2fc673b15196be26e3810302514d37d4c7a1e046

Reply:
>That's said, it's imho already much, much better than independent parameters. We could keep this action (but renamed), because what I feel is missing is the isEstimatedFeeLoading indicator. Now you don't really know if the fee is null because it's supposed to be null (due to some other things missing) or if it's actually loading.

I'm  afraid that i'm not following you on this one. Anyway i have changed the `setEstimateFee(null, null)` as you recommended so maybe this comment got outdated already.

Reply:
In a nutshell , i have removed the loading status and when call starts is calling:  setEstimateFee(null, null)

Reply:
So you seem to assume that the `async` call we are making is instant to the user (i.e. below 200ms). This is true on local development, but is not true in general.

I think we should have a loading state flag for every asynchronously updated parameter, so that we can display this to the user.
Now you are setting the fee to `null`- that's cool, the user will see that it's not there, however on the data model level you can't distinguish these two situations:
1. The fee is not there, because the input is somehow invalid and it will never be updated.
2. The fee is not there, because it's being calculated - the user should stay still because it will be updated shortly.

I really recommend testing the UI after enabling network throttling in the Chrome Dev Tools. It's not only about people on slow connections, but long term it's also about light(-er) clients that might take a while to process complex queries (like fee estimation).

Reply:
@tomusdrw  i would like to clarify this before merging anything. I'm feeling quite confused now about the loaders states because in previous comments it was requested to not use a separate loader variable in the reducer:

https://github.com/paritytech/parity-bridges-ui/pull/186#discussion_r650798957

https://github.com/paritytech/parity-bridges-ui/pull/186#discussion_r650801306

https://github.com/paritytech/parity-bridges-ui/pull/186#discussion_r654432343

My first approach was to use a dedicated variable for loading states as all other libraries / code do , but based on the discussions shared above i changed the code. If you asked me , my first and preferred approach was always to use dedicated loading variables for every call ( and in this case it will be in the reducer since it is a global value )

Reply:
What i can do is to put back specific loading variables for each action but make the dispatch function to be in charge of updating for example: 

`const setEstimatedFee = (estimatedFeeError: string | null, estimatingFeeLoading: boolean, estimatedFee: string | null) => {
  return {
    payload: { estimatedFee, estimatedFeeError, estimatingFeeLoading },
    type: TransactionActionTypes.SET_ESTIMATED_FEE
  };
};`

then on the `useDispatchGenericCall` i could use:

`  const execute = useCallback(
    async (...params: any[]) => {
      let data;
      let error;
      try {
        dispatch(null, true, null);
        data = await call(...params);
        dispatch(null, false, data);

      } catch (e) {
        error = e;
        dispatch(e, false,null);
      }
      return { data, error };
    },
    [call, dispatch]
  );`

Reply:
I think you are confusing two things:
1. Can we display the fee - i.e. is there a meaningful value in `estimateFee` variable.
2. Is the fee currently being loaded - i.e. `isFeeLoading === true`.

In the linked comments you've been using `isFeeLoading`:
1. either together with `estimateFee` check, which is superfluous if we know that every time `isFeeLoading === true` then `estimateFee === null`.
2. or incorrectly instead of `estimateFee === null` check (i.e. `estimateFee` can be `null` not only when it's loading).

Basically you want to prevent something from happening if the fee is not there (be it either display, or kicking off some extra calculations). Note that the loading state is not really relevant, cause it's only one case when fee is `null`, maybe there was some other error (i.e. missing parameters, API not available, etc) that makes the fee unavailable even if it's not loading.

The `isFeeLoading` variable should be used to display loading indicator to the user in case we know that the fee is being refreshed, but it should be ONLY used for that purpose - it's a pure user experience related state variable, everywhere else you should be using `estimateFee === null` cause it covers more broad spectrum of cases (missing data, errors, etc).


Also note that I was against introducing a separate `action` to update `isFeeLoading`, but not about having this value at all. Basically because the value of this variable is tightly coupled with other fee-related state variables I was suggesting that we should have two actions:
1. We are starting to make an asynchronous request to calculate the fee. (`startEstimateFee` or sth like that)
2. The asynchronous request to calculate the fee has finished (either error, timeout, or a success) (`finishEstimateFee` or sth like that)

And the first action should be responsible for `null`ifying the values and setting `isFeeLoading` to `true`, while the second action should be responsbile for flipping `isFeeLoading` back to false and setting either error or successful result to the state.

Reply:
> What i can do is to put back specific loading variables for each action but make the dispatch function to be in charge of updating for example:

By adding `isFeeLoading` parameter to the action creator you are exposing the internal details of state management to the component level and risk state inconsistency (what if a component sets `data` to something not `null`, but sets `loading` to `true`?).

Making reducer actions more "domain specific" is the way to go - that's the main distinction between regular `useState`, where you basically have no central state management, but you expose all details to the component level - which is actually much better (less boilerplate) in case of simple state. Reducers really shine in complex state management, but in such case you also need to have domain specific actions, and not simply `setters` for values that the reducer state holds.

Reply:
But as I said earlier in some comment, I feel long-term consistency will be much more valueable than simply addressing this comments in one PR. Like if you fundamentally don't think that domain-specific actions are easier to reason about and you prefer the setter approach, let's do it consistently everywhere, I'm not going to die on this hill, but just telling you from my experience that state consistency and simplicity of interactions makes the code easier to reason about and debug.